lib/core/src/Cardano/Wallet/Logging.hs

```diff
@@ -148,86 +160,187 @@ stdoutTextTracer = Tracer $ liftIO . B8.putStrLn . T.encodeUtf8 . toText
 -------------------------------------------------------------------------------}
 
 -- | Run an 'ExceptT' action, then trace its result, all in one step.
+-- This is a more basic version of 'resultTracer'.
 traceWithExceptT :: Monad m => Tracer m (Either e a) -> ExceptT e m a -> ExceptT e m a
 traceWithExceptT tr (ExceptT action) = ExceptT $ do
     res <- action
     traceWith tr res
     pure res
 
--- | Convert an IO tracer to a 'm' tracer.
-unliftIOTracer :: MonadIO m => Tracer IO a -> Tracer m a
-unliftIOTracer = natTracer liftIO
+-- | Log around an 'ExceptT' action. The result of the action is captured as an
+-- 'Either' in the log message. Other unexpected exceptions are captured in the
+-- 'BracketLog''.
+traceResult
+    :: MonadUnliftIO m
+    => Tracer m (BracketLog' (Either e r))
+    -> ExceptT e m r
+    -> ExceptT e m r
+traceResult tr = ExceptT . bracketTracer' id tr . runExceptT
+
+-- | Format a tracer message from 'traceResult' as multiline text.
+formatResultMsg
+    :: (Show e, IsList t, Item t ~ (Text, v), Buildable v, Buildable r)
+    => Text
+    -- ^ Function name.
+    -> t
+    -- ^ Input parameters.
+    -> BracketLog' (Either e r)
+    -- ^ Logging around function.
+    -> Builder
+formatResultMsg = formatResultMsgWith (nameF "ERROR" . build . show) build
+
+-- | Same as 'formatResultMsg', but accepts result formatters as parameters.
+formatResultMsgWith
+    :: (IsList t, Item t ~ (Text, v), Buildable v)
+    => (e -> Builder)
+    -- ^ Error message formatter
+    -> (r -> Builder)
+    -- ^ Result formatter
+    -> Text
+    -- ^ Function name.
+    -> t
+    -- ^ Input parameters.
+    -> BracketLog' (Either e r)
+    -- ^ Logging around function.
+    -> Builder
+formatResultMsgWith err fmt title params b = nameF (build title) $ blockListF
+    [ nameF "inputs" (blockMapF params)
+    , buildBracketLog (either err fmt) b
+    ]
+
+-- | A good default mapping of message severities for 'traceResult'.
+resultSeverity :: Severity -> BracketLog' (Either e r) -> Severity
+resultSeverity base = \case
+    BracketStart -> base
+    BracketFinish (Left _) -> Error
+    BracketFinish (Right _) -> base
+    BracketException _ -> Error
+    BracketAsyncException _ -> base
```

Comment:
Hmm, why does this return `base` instead of error? What about the asynchronous nature of the exception encourages this distinction?